File: Lv.2/괄호 변환/solution.py
# https://school.programmers.co.kr/learn/courses/30/lessons/60058
class Stack:
    def __init__(self):
        self.s = []
        
    def push(self, b):
        self.s.append(b)
        
        slen = len(self.s)
        if slen > 1 and self.s[slen - 1] == ')' and self.s[slen - 2] == '(':
            self.s.pop()
            self.s.pop()        
        
    def is_empty(self):
        if self.s:
            return False
        else:
            return True

def solution(p):
    if p == '':
        return ''
    
    oc = 0
    cc = 0
    u = ''
    v = ''
    
    for i, c in enumerate(p):
        if c == '(':
            oc += 1
        elif c == ')':
            cc += 1
            
        if oc == cc:
            u = p[:i+1]
            v = p[i+1:]
            break
            
    s = Stack()
    for c in u:
        s.push(c)
            
    if s.is_empty():
        return u + solution(v)
    else:
        e = '(' + solution(v) + ')'
        
        if u:
            u = list(u)[1:]
            u.pop()
        else:
            u = []
        
        for i, c in enumerate(u):
            if c == ')':
                u[i] = '('
            elif c == '(':
                u[i] = ')'
                
        u = ''.join(u)
                
        return e + u

# pythonic한 방법
def solution(p):
    if p=='': return p
    r=True; c=0
    for i in range(len(p)):
        if p[i]=='(': c-=1
        else: c+=1
        if c>0: r=False
        if c==0:
            if r:
                return p[:i+1]+solution(p[i+1:])
            else:
                return '('+solution(p[i+1:])+')'+''.join(list(map(lambda x:'(' if x==')' else ')',p[1:i]) ))

File: Lv.2/괄호 변환/solution.js
// https://school.programmers.co.kr/learn/courses/30/lessons/60058?language=javascript
class Stack {
  s = [];

  push = (b) => {
    this.s.push(b);
    const s_len = this.s.length;
    if (s_len > 1 && this.s[s_len - 1] === ")" && this.s[s_len - 2] === "(") {
      this.s.pop();
      this.s.pop();
    }
  };

  is_empty = () => {
    return this.s.length ? false : true;
  };
}

function solution(p) {
  if (p === "") {
    return "";
  }

  let [oc, cc, u, v] = [0, 0, "", ""];

  for (let i = 0; i < p.length; i += 1) {
    if (p[i] === "(") {
      oc += 1;
    } else if (p[i] == ")") {
      cc += 1;
    }

    if (oc === cc) {
      u = p.slice(0, i + 1);
      v = p.slice(i + 1);
      break;
    }
  }

  const stack = new Stack();
  [...u].forEach((e) => {
    stack.push(e);
  });

  if (stack.is_empty()) {
    return u + solution(v);
  } else {
    const e = "(" + solution(v) + ")";

    if (u.length) {
      u = [...u.slice(1, u.length - 1)];
    } else {
      u = [];
    }

    for (let i = 0; i < u.length; i += 1) {
      if (u[i] === ")") {
        u[i] = "(";
      } else if (u[i] === "(") {
        u[i] = ")";
      }
    }

    return e + u.join("");
  }
}
